### PR TITLE
feat: #87 add_InProgress_component

### PR DESCRIPTION
**Pull request by barrydilan**
Sep-22-2023 15-45-42

**Description:** 

As of this moment, inProgress component is a progress bar line that is rendered in the head of a table on each render using setTimeout function. As far as i understand, the  **DayReportService.onValueFromRealTimeDb(setData);** inside Table.jsx returns an array of arrays, and only then the Table is created, so I don`t understand how to check for individual cell loading status. However, the inProgressComponent can be rendered inside any cell with position: relative property.

### CODE REVIEW

File: mw-webapp/src/component/inProgress/InProgress.module.scss [hunk 1]
```diff
@@ -0,0 +1,15 @@
+.container {
+  width: 0%;
+  position: absolute;
+  bottom: 0;
+  left: 0;
+  height: 2px;
+  background-color: #007bff;
```
Comment (Ekaterina1994) [inline mw-webapp/src/component/inProgress/InProgress.module.scss:7]: use global variables (folder styles)

File: mw-webapp/src/component/table/Table.tsx [hunk 2]
```diff
@@ -1,15 +1,25 @@
 import {useEffect, useState} from "react";
-import {flexRender, getCoreRowModel, useReactTable} from "@tanstack/react-table";
+import InProgress from "../inProgress/InProgress";
```
Comment (Ekaterina1994) [inline mw-webapp/src/component/table/Table.tsx:2]: absolute paths

File: mw-webapp/src/component/table/Table.tsx [hunk 3]
```diff
@@ -1,15 +1,25 @@
...
 
 export const Table = () => {
   const [data, setData] = useState<DayReport[]>([]);
+  const [loading, setLoading] = useState(true);
 
   useEffect(() => {
     DayReportService.onValueFromRealTimeDb(setData);
+    setTimeout(() => {
+      setLoading(false);
+    // eslint-disable-next-line no-magic-numbers
```
Comment (Ekaterina1994) [inline mw-webapp/src/component/table/Table.tsx:21]: don't use disable of eslint rules. If we will use disable then why we need eslint))

**Comment by tritonJS826** [comment]
Hey Cool) Let me specify feature requirements) We want to show the user something when a network request is in progress, a kind of loader but without blocking UI. We chose the progress bar. It should appear when any request sent and disappeared when all request finished. We organize our network function with classes. I propose to use method decorator. It will trigger progressbar appearence when some of decorated method will be invoked. Look in direction of this pakcage, investigate, maybe you will find something better: https://www.npmjs.com/package/@tanem/react-nprogress

File: mw-webapp/src/component/table/Table.tsx [hunk 4]
```diff
@@ -1,15 +1,25 @@
...
+  useReactTable,
+} from "@tanstack/react-table";
 import {columns} from "src/component/table/columns";
 import {DayReport} from "src/model/businessModel/DayReport";
 import {DayReportService} from "src/service/DayReportService";
 import styles from "src/component/table/Table.module.scss";
 
 export const Table = () => {
   const [data, setData] = useState<DayReport[]>([]);
+  const [loading, setLoading] = useState(true);
```
Comment (Ekaterina1994) [inline mw-webapp/src/component/table/Table.tsx:15]: usully if value is boolean we  name variable is... (isLoading)

File: mw-webapp/src/component/inProgress/InProgress.tsx [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import styles from "src/component/inProgress/InProgress.module.scss";
+
+const InProgress = () => {
+  return (
+    <div className={styles.container} />
+  );
+};
+
+export default InProgress;
```
Comment (tritonJS826) [inline mw-webapp/src/component/inProgress/InProgress.tsx:9]: no default exports, please adjust this rule in eslint for everybody

File: mw-webapp/src/component/inProgress/InProgress.tsx [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import styles from "src/component/inProgress/InProgress.module.scss";
+
+const InProgress = () => {
```
Comment (tritonJS826) [inline mw-webapp/src/component/inProgress/InProgress.tsx:3]: add docs for all public members

**Review by tritonJS826 — CHANGES_REQUESTED** [review]

File: mw-webapp/src/component/table/Table.tsx [hunk 7]
```diff
@@ -1,15 +1,25 @@
 import {useEffect, useState} from "react";
```
Comment (Ekaterina1994) [inline mw-webapp/src/component/table/Table.tsx:1]: You can don't add component to table. Task is just create component. If you will use it in other files - there will be more mistakes and conflicts

File: mw-webapp/src/component/table/columns.tsx [hunk 8]
```diff
@@ -1,4 +1,8 @@
-import {CellContext, ColumnDef, createColumnHelper} from "@tanstack/react-table";
```
Comment (Ekaterina1994) [inline mw-webapp/src/component/table/columns.tsx:1]: What was changed in this file? Didn't found(

**Review by Ekaterina1994 — CHANGES_REQUESTED** [review]

**Comment by barrydilan** [comment]
> Hey Cool, thanks) Let me specify feature requirements) We want to show the user something when a network request is in progress, a kind of loader but without blocking UI. We chose the progress bar. It should appear when any request is sent and disappear when all requests are finished. We organize our network function with classes. I propose to use the method decorator. It will trigger a progress bar appearance when some of the decorated methods are invoked. Look in the direction of this package, investigate, and maybe you will find something better: https://www.npmjs.com/package/@radix-ui/react-progress BTW I think it will be better to place progress bare somewhere near the header instead of some table)

Thanks for the clarification, I must have misunderstood the task the first time

File: mw-webapp/src/component/table/columns.tsx (hunk 8 above)
Comment (barrydilan) [reply mw-webapp/src/component/table/columns.tsx:1]: perhaps formatting

File: mw-webapp/src/component/inProgress/loading.tsx
Comment (barrydilan) [inline mw-webapp/src/component/inProgress/loading.tsx:1]: I cannot figure out how to render component.
As far as I understand, I need to manipulate some sort of state but I cannot do it outside of a react component.
Am I moving in the right direction?

Comment (Ekaterina1994) [reply mw-webapp/src/component/inProgress/loading.tsx:1]: You can manipulate of useState in functional components. And then use useEffect for change value in InProgress component? Did you tried this way?

Comment (barrydilan) [reply mw-webapp/src/component/inProgress/loading.tsx:1]: Yes, what I mean is that Im trying to use a decorator   
@withLoading() on a onValueFromRealTimeDb method like [tritonJS826](https://github.com/tritonJS826) suggested, so that when the request to db is fired it will trigger a progress bar appearance.
The problem is that decorator is used for modifying the logic of methods or properties, not for rendering UI components, so in order to render JSX I need to handle that logic separately. 

export default function withLoading() {
  return function WithLoadingHOC(
....
const originalMethod = descriptor.value;
    descriptor.value = async function (...args: unknown[]) {
      try {
        // SetLoading(true); <======= manipulating state
        const result = await originalMethod.apply(this, args);
        return result;
      } finally {
        // SetLoading(false);
      }
    };
This is what Im trying to do in theory

Comment (tritonJS826) [reply mw-webapp/src/component/inProgress/loading.tsx:1]: You can use just high ordered function, or just callback for the first version of it will be hard to implement with decisions. As I understand , the main question here - how to integrate react with outside methods. There are many variants from using some state manager with injecting store , just for example redux (https://stackoverflow.com/questions/65789215/how-to-access-react-redux-store-outside-react-component-in-redux-toolkit), to using useSyncExternalStore ( https://react.dev/reference/react/useSyncExternalStore).

File: .DS_Store
Comment (Ekaterina1994) [inline .DS_Store:1]: Do we need this file on github??

Comment (Ekaterina1994) [reply .DS_Store:1]: And if yes - then what is it?

File: mw-webapp/.eslintrc [hunk 9]
```diff
@@ -18,8 +18,9 @@
     // "project": ["**/tsconfig.json", "**/vite.config.ts"] // путь для сопоставления файлов tsconfig относительно parserOptions.tsconfigRootDir, ​​или текущего рабочего каталога, если он не указан
     "project": true // указывает на поиск ближайшего tsconfig.json для каждого исходного файла
   },
-  "plugins": ["react", "@typescript-eslint", "simple-import-sort"],
+  "plugins": ["react", "@typescript-eslint", "simple-import-sort", "import"],
   "rules": {
+    "import/no-default-export": "off",
```
Comment (Ekaterina1994) [inline mw-webapp/.eslintrc:23]: Can't we use https://eslint.org/docs/latest/rules/no-restricted-exports instead of add new dependency?
If you think that we can't use it then explain please why?

File: mw-webapp/package.json [hunk 10]
```diff
@@ -72,11 +72,12 @@
     "@typescript-eslint/parser": "^6.4.0",
     "@vitejs/plugin-react": "^4.0.4",
     "eslint": "^8.47.0",
+    "eslint-plugin-import": "^2.28.1",
```
Comment (Ekaterina1994) [inline mw-webapp/package.json:75]: Can't we use https://eslint.org/docs/latest/rules/no-restricted-exports instead of add new dependency?
If you think that we can't use it then explain please why?

File: mw-webapp/src/component/inProgress/InProgress.module.scss [hunk 11]
```diff
@@ -0,0 +1,16 @@
+.container {
+  width: 0%;
+  position: absolute;
+  top: 0;
+  left: 0;
+  height: 4px;
+  background-color: white;
+  // background-color: $progressBar-background;
+  animation: progress 1s linear forwards;
```
Comment (Ekaterina1994) [inline mw-webapp/src/component/inProgress/InProgress.module.scss:9]: Delete any unused comments

File: mw-webapp/src/component/inProgress/InProgress.module.scss [hunk 12]
```diff
@@ -0,0 +1,16 @@
+.container {
+  width: 0%;
+  position: absolute;
+  top: 0;
+  left: 0;
+  height: 4px;
+  background-color: white;
```
Comment (Ekaterina1994) [inline mw-webapp/src/component/inProgress/InProgress.module.scss:7]: You can dont add styles or if you want to use it then we use global vars for colors, fonts paddings, etc
So or delete styles or move somw prperties to global

File: mw-webapp/src/component/inProgress/InProgress.tsx (hunk 6 above)
Comment (Ekaterina1994) [reply mw-webapp/src/component/inProgress/InProgress.tsx:3]: Sorry but don't see docs. Why you resolved this conversation?

File: mw-webapp/src/component/inProgress/InProgress.tsx (hunk 5 above)
Comment (Ekaterina1994) [reply mw-webapp/src/component/inProgress/InProgress.tsx:9]: We don't use default export/ Maybe some changes was not pushed?

File: mw-webapp/src/component/table/LoadingContext.tsx [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+type LoadingContextType = {
+  isLoading: boolean;
+  setLoading: React.Dispatch<React.SetStateAction<boolean>>;
+};
+
+const LoadingContext = createContext<LoadingContextType | undefined>(undefined);
+
+type LoadingProviderProps = {
+  children: ReactNode;
+};
```
Comment (Ekaterina1994) [inline mw-webapp/src/component/table/LoadingContext.tsx:12]: We don't use children in props/ You can use PropsWithChildren

File: mw-webapp/src/component/table/columns.tsx [hunk 14]
```diff
@@ -11,121 +15,129 @@ const DEFAULT_SUMMARY_TIME = 0;
...
-  return (
-    (!arrayItem) ?
-      <div />
-      :
-    //TODO: task #65 use flag instead of first index
-      <div key={index}>
-        <div className={arrayItem[INDEX_OF_CHECK_MARK] === "✓" ? styles.completed : styles.notCompleted}>
-          {arrayItem}
-        </div>
+  return !arrayItem ? (
```
Comment (Ekaterina1994) [inline mw-webapp/src/component/table/columns.tsx:32]: Why these changes? Why you deleted "(" . Explain please

File: mw-webapp/src/component/table/LoadingContext.tsx [hunk 15]
```diff
@@ -0,0 +1,30 @@
+import React, {createContext, ReactNode, useContext, useState} from "react";
+
+type LoadingContextType = {
+  isLoading: boolean;
+  setLoading: React.Dispatch<React.SetStateAction<boolean>>;
+};
+
+const LoadingContext = createContext<LoadingContextType | undefined>(undefined);
```
Comment (Ekaterina1994) [inline mw-webapp/src/component/table/LoadingContext.tsx:8]: Why you decide to use useContext?

File: mw-webapp/src/component/table/Table.tsx [hunk 16]
```diff
@@ -1,15 +1,24 @@
...
-import {flexRender, getCoreRowModel, useReactTable} from "@tanstack/react-table";
+import InProgress from "../inProgress/InProgress";
+import {useLoading} from "./LoadingContext";
+import {
+  flexRender,
+  getCoreRowModel,
+  useReactTable,
+} from "@tanstack/react-table";
```
Comment (Ekaterina1994) [inline mw-webapp/src/component/table/Table.tsx:8]: We use absolute path

You shouldn't use component inside table component/ Your task is create InProgress component/ You can don't change table , columns and other file

**Review by Ekaterina1994 — CHANGES_REQUESTED** [review]

**Review by tritonJS826 — CHANGES_REQUESTED** [review]
Please solve previous issues

File: .DS_Store
Comment (barrydilan) [reply .DS_Store:1]: gitignor updated

File: mw-webapp/src/component/table/columns.tsx (hunk 14 above)
Comment (barrydilan) [reply mw-webapp/src/component/table/columns.tsx]: Must have been done automatically by prettier

File: mw-webapp/.eslintrc (hunk 9 above)
Comment (barrydilan) [reply mw-webapp/.eslintrc]: updated

File: mw-webapp/src/component/inProgress/InProgress.module.scss (hunk 11 above)
Comment (barrydilan) [reply mw-webapp/src/component/inProgress/InProgress.module.scss:8]: deleted

File: mw-webapp/src/component/table/Table.tsx (hunk 16 above)
Comment (barrydilan) [reply mw-webapp/src/component/table/Table.tsx:6]: all modifications to Table deleted

File: mw-webapp/src/component/table/LoadingContext.tsx (hunk 15 above)
Comment (barrydilan) [reply mw-webapp/src/component/table/LoadingContext.tsx:8]: I didnt want to use any external state manager

File: mw-webapp/src/component/inProgress/InProgress.tsx (hunk 6 above)
Comment (barrydilan) [reply mw-webapp/src/component/inProgress/InProgress.tsx]: sorry, my bad, missed during conflicts resolving.

File: mw-webapp/.eslintrc [hunk 17]
```diff
@@ -29,6 +29,7 @@
         ]
       }
     ],
+    "no-restricted-exports": ["error", { "restrictedNamedExports": ["default"] }],
```
Comment (Ekaterina1994) [inline mw-webapp/.eslintrc:32]: "no-restricted-exports": ["error", {"restrictDefaultExports":  [direct]"}]

File: mw-webapp/src/component/inProgress/InProgress.tsx [hunk 18]
```diff
@@ -0,0 +1,12 @@
+import styles from "src/component/inProgress/InProgress.module.scss";
+
+/**
+ * Represents a component that renders a Progress Bar
+ * @component
+ */
+
+export const InProgress = () => {
```
Comment (Ekaterina1994) [inline mw-webapp/src/component/inProgress/InProgress.tsx:8]: Delete empty line

File: mw-webapp/src/component/table/Table.tsx [hunk 19]
```diff
@@ -23,11 +26,11 @@ export const Table = () => {
     getCoreRowModel: getCoreRowModel(),
   });
 
-  return (
+  return (<>
```
Comment (Ekaterina1994) [inline mw-webapp/src/component/table/Table.tsx:29]: Delete <>

File: mw-webapp/src/styles/_variables.scss [hunk 20]
```diff
@@ -12,4 +12,7 @@ $link-active: rgb(255, 149, 0);
 $header-background: rgb(85, 122, 149);
 
 // Title
-$title-padding: 10px;
\ No newline at end of file
+$title-padding: 10px;
+
+//ProgressBar
```
Comment (Ekaterina1994) [inline mw-webapp/src/styles/_variables.scss:17]: Add space before Start of comment

**Review by Ekaterina1994 — CHANGES_REQUESTED** [review]

**Review by Ekaterina1994 — COMMENTED** [review]
Please resolve conflicts before next rereview and don't resolve conversations/ Conversations should be closed by person who open it)

File: mw-webapp/.eslintrc [hunk 21]
```diff
@@ -29,6 +29,7 @@
         ]
       }
     ],
+    "no-restricted-exports": ["error", {"restrictedNamedExports": ["default"]}],
     "capitalized-comments": [
```
Comment (Ekaterina1994) [inline mw-webapp/.eslintrc:33]: why not "no-restricted-exports": ["error", {"restrictDefaultExports": [direct]"}]
Can you please explain fifference bwtween these 2 rules

File: mw-webapp/src/pages/mainPage/MainPage.tsx [hunk 22]
```diff
@@ -1,3 +1,4 @@
+import {LoadingProvider} from "src/component/table/LoadingContext";
 import {Table} from "src/component/table/Table";
```
Comment (Ekaterina1994) [inline mw-webapp/src/pages/mainPage/MainPage.tsx:2]: Do you use LoadingProvider in this file? Sorry but dont see in code

**Review by Ekaterina1994 — CHANGES_REQUESTED** [review]

File: mw-webapp/.eslintrc (hunk 21 above)
Comment (barrydilan) [reply mw-webapp/.eslintrc:33]: i think they are interchangable as per doc. Changed to "no-restricted-exports": ["error", {"restrictDefaultExports": [direct]"}] anyway.

File: mw-webapp/src/pages/mainPage/MainPage.tsx (hunk 22 above)
Comment (barrydilan) [reply mw-webapp/src/pages/mainPage/MainPage.tsx]: forget to delete import after cleaning up component

File: .gitignore [hunk 23]
```diff
@@ -133,4 +133,8 @@ dist
 .vscode
 
 # autosave draw.io
+.$models.drawio.bkp
+
+# misc
+.DS_Store
```
Comment (tritonJS826) [inline .gitignore:139]: Looks like it is something related to mac. Please add short comment to clarify this for other developers

File: mw-webapp/vite.config.ts [hunk 24]
```diff
@@ -3,6 +3,7 @@ import {defineConfig, loadEnv} from "vite";
 import eslint from "vite-plugin-eslint";
 import viteTsconfigPaths from "vite-tsconfig-paths";
 
+// eslint-disable-next-line no-restricted-exports
 export default defineConfig(() => {
```
Comment (tritonJS826) [inline mw-webapp/vite.config.ts:7]: could you exalude this file from linting and remove comment please?

**Review by tritonJS826 — APPROVED** [review]
Minor changes required

**Review by Ekaterina1994 — COMMENTED** [review]
Reslve conflicts for rereview pls

File: mw-webapp/vite.config.ts (hunk 24 above)
Comment (barrydilan) [reply mw-webapp/vite.config.ts]: done

File: .gitignore (hunk 23 above)
Comment (barrydilan) [reply .gitignore:139]: moved under #macOS section

**Review by Ekaterina1994 — APPROVED** [review]
